samples/getting-started/simple-algorithms/HiddenShift.qs

```diff
@@ -0,0 +1,196 @@
+    // a diagonal operator that maps |x〉 -> (-1)^{IP(x)} |x〉, where x stands for
+    // x = (x_0, ..., x_{n-1}) and all the x_i are binary. The IP function is
+    // defined as IP(y, z) = y_0 z_0 + y_1 z_1 + ... y_{u-1} z_{u-1} where
+    // y = (y_0, ..., y_{u-1}) and z =  (z_0, ..., z_{u-1}) are two bit
+    // vectors of length u. Notice that the function IP is a Boolean function
+    // on n = 2u bits. IP is a special case of a so-called 'bent' function.
+    // These are functions for which the Walsh-Hadamard transform is perfectly
+    // flat (in absolute value). Because of this flatness, the Walsh-Hadamard
+    // spectrum of any bent function defines a +1/-1 function, i.e., gives
+    // rise to another Boolean function, called the 'dual bent function'.
+    // What is more, for the case of the IP function it can be shown that IP
+    // is equal to its own dual bent function, a fact that is exploited in
+    // the present test case.
+
+    // Notice that a diagonal operator implementing IP between 2 variables
+    // y_0 and z_0 is nothing but the AND function between those variables, i.e.,
+    // in phase encoding it is computed by a Controlled-Z gate. Extending this
+    // to an XOR of the AND of more variables, as required in the definition of
+    // the IP function can then be accomplished by applying several Controlled-Z
+    // gates between the respective inputs.
+    internal operation ApplyInnerProductBentFunction(u : Int, qs : Qubit[]) : Unit {
+
+        if (Length(qs) != 2 * u) {
+            fail "Length of qs must be twice the value of u";
+        }
+
+        let xs = qs[0 .. u - 1];
+        let ys = qs[u .. 2 * u - 1];
+
+        for (idx in 0 .. u - 1) {
+            Controlled Z([xs[idx]], ys[idx]);
+        }
+    }
+
+
+    // Again, using partial application we create a function which for a given bit
+    // size u constructs the IP Boolean function on 2u qubits, computed into the phase.
+    function InnerProductBentFunction(u : Int) : (Qubit[] => Unit) {
+        return ApplyInnerProductBentFunction(u, _);
+    }
+
+
+    // To instantiate the hidden shift problem we need another function g which is
+    // related to IP via g(x) = IP(x + s), i.e., we have to shift the argument of
+    // the IP function by a given shift. Notice that the '+' operation here is the
+    // Boolean addition, i.e., a bit-wise operation. Notice further, that in
+    // general a diagonal operation |x〉 -> (-1)^{f(x)} can be turned into a shifted
+    // version by applying a bit flip to the |x〉 register first, then applying the
+    // diagonal operation, and then undoing the bit flips to the |x〉 register. We
+    // use this principle to define shifted versions of the IP operation.
+    internal operation ApplyShiftedInnerProductBentFunction(shift : Bool[], u : Int, qs : Qubit[]) : Unit {
+        let n = 2 * u;
+
+        if (Length(shift) != n or Length(qs) != n) {
```

Comment:
This could be split into two `EqualityFactI` assertions which can make it easier to spot which of the arrays had the wrong size.